src/mpi/init/mpir_init.c

```diff
@@ -114,21 +114,6 @@ int MPII_Init_thread(int *argc, char ***argv, int user_required, int *provided,
 
     MPL_initlock_lock(&MPIR_init_lock);
 
-    if (!is_world_model) {
-        *p_session_ptr = (MPIR_Session *) MPIR_Handle_obj_alloc(&MPIR_Session_mem);
-        MPIR_ERR_CHKHANDLEMEM(*p_session_ptr);
```

Comment:
Looking at your other comment, I'm not sure if this comment still requires an answer or not. :)

Are there still open concerns on your side?

Reply:
This still need be addressed. `MPIR_Handle_obj_alloc` need be inside a critical section. Otherwise, there will be race condition when two threads concurrently calls `MPI_Session_init`.

Reply:
I see. I revised the solution to keep the session's object creation within `MPII_Init_thread` and only do the parsing of the info object for the `thread_level` parameter before.